### ISSUE TITLE
dd-trace makes jest transform cache keys unstable

### ISSUE THREAD

**Issue by abejfehr** [DataDog/dd-trace-js#3389]
**Expected behaviour**

Running a single suite of tests 2 times in a row in CI should result in:
1. the first run taking a long(ish) amount of time
2. the second run taking a significantly shorter amount of time, since the jest transform cache is warm

Additionally, I should be able to persist the jest transform cache for a future run to reduce test times.

**Actual behaviour**

Running a single suite of tests 2 times in a row in CI take equally long.

Even with a persisted jest transform cache there are no cache hits.

**Steps to reproduce**

The steps to reproduce are already outlined in the "Expected behaviour" section above.

**Environment**

* **Operation system:** linux (in CI)
* **Node.js version:** v16.20.1
* **Tracer version:** N/A
* **Agent version:** N/A
* **Relevant library versions:** 4.2.0 (although I'm not sure which others are affected)

**Comment by abejfehr** [comment]
Just a brief explanation:

1. Jest generates the cacheKey for the transform cache by concatenating many things together, for example in `babel-jest` the configString is part of the concatenation ([link](https://github.com/jestjs/jest/blob/70b17d2e3d2bf40dd27fecdcdd06c159d26a7c6c/packages/babel-jest/src/index.ts#L90))
2. Jest provides the configString here in their ScriptTransformer, which is just a stringified version of the Jest config ([link](https://github.com/jestjs/jest/blob/main/packages/jest-transform/src/ScriptTransformer.ts#L96))
3. The Jest config is made to include several unstable (different between runs) values by dd-trace ([link](https://github.com/DataDog/dd-trace-js/blob/master/packages/datadog-plugin-jest/src/index.js#L81-L85))

<img width="1691" alt="image" src="https://github.com/DataDog/dd-trace-js/assets/3917428/74bd8cb6-6935-4b1e-8ee8-92b9d1b3cd85">

Those differences in `configString` cause the cache keys to miss, which causes tests to take longer

**Comment by juan-fernandez** [comment]
hey @abejfehr! Thanks for the report and investigation. 

We're aware of this issue but thought it wouldn't be a problem, as we understand that this jest cache is usually used for local runs, not CI runs, where `dd-trace` would be active. 

Could you give me more details about how you're using jest cache for your CI pipeline?

**Comment by abejfehr** [comment]
All our front end code is in a monorepo and we have a _lot_ of tests, so it's a huge help for runtime to persist the Jest transform cache.

We persist it between runs with [actions/cache](https://github.com/actions/cache) and prune the oldest entries before persisting to keep it a reasonable size

**Comment by juan-fernandez** [comment]
> All our front end code is in a monorepo and we have a _lot_ of tests, so it's a huge help for runtime to persist the jest transform cache.
> 
> We persist it between runs with [actions/cache](https://github.com/actions/cache) and prune the oldest entries before persisting to keep it a reasonable size

thanks for the context! Could you share part of the logic in your github action here? You can remove anything you think it's sensitive, but this would greatly help me test this 😄

**Comment by abejfehr** [comment]
If you'd just like to test this, you can just run tests twice in a row in a github action, like so:

```
- name: Unit Tests
  run: yarn test
  env:
    DD_CIVISIBILITY_AGENTLESS_ENABLED: true
    DD_API_KEY: ${{ secrets.DD_API_KEY }}
    NODE_OPTIONS: '-r dd-trace/ci/init'
    DD_ENV: ci
    DD_SERVICE: front-end-monorepo

- name: Unit Tests Again
  run: yarn test
  env:
    DD_CIVISIBILITY_AGENTLESS_ENABLED: true
    DD_API_KEY: ${{ secrets.DD_API_KEY }}
    NODE_OPTIONS: '-r dd-trace/ci/init'
    DD_ENV: ci
    DD_SERVICE: front-end-monorepo
```

If you do this, you'll notice that both tests take the same amount of time (which is not good).

You can actually get around this though if you'd like to test our patch. Our workaround is to change [this line](https://github.com/jestjs/jest/blob/2499cf2a3a75824fd0c9b454baa95fc75929f086/packages/jest-transform/src/ScriptTransformer.ts#L96) in the `@jest/transform` package from:

```
const configString = stableStringify(this._config);
```

to

```
const stableConfig = { ...this._config };
delete stableConfig.testEnvironmentOptions
const configString = stableStringify(stableConfig);
```

so that we don't get `testEnvironmentOptions` in the cacheKey

**Comment by juan-fernandez** [comment]
> If you'd like to test this, you can just run tests twice in a row in a github action, like so:
> 
> ```
> - name: Unit Tests
>   run: yarn test
>   env:
>     DD_CIVISIBILITY_AGENTLESS_ENABLED: true
>     DD_API_KEY: ${{ secrets.DD_API_KEY }}
>     NODE_OPTIONS: '-r dd-trace/ci/init'
>     DD_ENV: ci
>     DD_SERVICE: my-monorepo
> 
> - name: Unit Tests Again
>   run: yarn test
>   env:
>     DD_CIVISIBILITY_AGENTLESS_ENABLED: true
>     DD_API_KEY: ${{ secrets.DD_API_KEY }}
>     NODE_OPTIONS: '-r dd-trace/ci/init'
>     DD_ENV: ci
>     DD_SERVICE: my-monorepo
> ```
> 
> If you do this, you'll notice that both tests take the same amount of time (which is not good).
> 
> You can actually get around this though if you'd like to test our patch. Our workaround is to change [this line](https://github.com/jestjs/jest/blob/2499cf2a3a75824fd0c9b454baa95fc75929f086/packages/jest-transform/src/ScriptTransformer.ts#L96) in the `@jest/transform` package from:
> 
> ```
> const configString = stableStringify(this._config);
> ```
> 
> to
> 
> ```
> const stableConfig = { ...this._config };
> delete stableConfig.testEnvironmentOptions;
> const configString = stableStringify(stableConfig);
> ```
> 
> so that we don't get `testEnvironmentOptions` in the cacheKey

thanks! That's very helpful. We might be able to do what you did in `@jest/transform` in the library. I'll keep you posted

**Comment by juan-fernandez** [comment]
hey! As an update, I made the cache work with this change: https://github.com/DataDog/dd-trace-js/compare/juan-fernandez/fix-cache-issues-with-jest, which is basically making it ignore the fields we add to `testEnvironmentOptions`, following your suggestion. I pushed these changes but I think I came up with a way not to use these fields at all, so we don't have to hack our way around the cache like this. I'll keep you updated 😄
